FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+                entity.NextJobId++;
+            }
+
+            try
+            {
+                await _azureJobInfoTableClient.UpdateEntityAsync(entity, entity.ETag, cancellationToken: cancellationToken);
+            }
+            catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.UpdateEntityPreconditionFailedErrorCode))
+            {
+                _logger.LogWarning(ex, "Update job id entity conflicts, will try again.");
+
+                // try to get job ids again
+                result = await GetIncrementalJobIds(queueType, count, cancellationToken);
+            }
+
+            return result;
+        }
+
+        private async Task<JobReverseIndexEntity> GetJobReverseIndexEntityByIdAsync(byte queueType, long jobId, CancellationToken cancellationToken)
+        {
+            var reversePartitionKey = AzureStorageKeyProvider.JobReverseIndexPartitionKey(queueType, jobId);
+            var reverseRowKey = AzureStorageKeyProvider.JobReverseIndexRowKey(queueType, jobId);
+            var reverseIndexResponse = await _azureJobInfoTableClient.GetEntityAsync<JobReverseIndexEntity>(reversePartitionKey, reverseRowKey, cancellationToken: cancellationToken);
+            return reverseIndexResponse.Value;
+        }
+
+        private static bool IsSpecifiedErrorCode(RequestFailedException exception, string expectedErrorCode) =>
+            string.Equals(exception.ErrorCode, expectedErrorCode, StringComparison.OrdinalIgnoreCase);
+
+        private static string FilterJobInfosByGroupId(byte queueType, long groupId) =>
+            $"PartitionKey eq '{AzureStorageKeyProvider.JobInfoPartitionKey(queueType, groupId)}' and RowKey ge '{groupId:D20}' and RowKey lt '{groupId + 1:D20}'";
+
+        private static IEnumerable<string> SelectPropertiesExceptDefinition()
+        {
+            var type = Type.GetType("Microsoft.Health.Fhir.Synapse.JobManagement.Models.AzureStorage.JobInfoEntity");
+            if (type == null)
+            {
+                throw new JobManagementException("Failed to get JobInfoEntity properties, the type is null.");
+            }
+
+            return type.GetProperties().Select(p => p.Name).Except(new List<string> { "Definition" });
+        }
+
+        /// <summary>
+        /// when cancel a job, always set the cancelRequest to true, and set its status to cancelled only when it's created,
+        /// for other cases, shouldCancel will be returned when keep alive, and jobHosting will cancel this job and set job to completed
+        /// </summary>
+        private static void CancelJobInternal(JobInfoEntity jobInfoEntity)
+        {
+            // set jobInfo entity's cancel request to true.
+            jobInfoEntity.CancelRequested = true;
+
+            // only set job status to cancelled when the job status is created.
+            if (jobInfoEntity.Status == (int)JobStatus.Created)
+            {
+                jobInfoEntity.Status = (int)JobStatus.Cancelled;
```

Comment:
For running jobs, we only set the `CancelRequest` in the entity to true, and when keeping alive this job, it will return `shouldCancel` to `jobHosting`, and `jobHosting` will cancel this job.